### PR TITLE
[4.0][Refactor][Breaking change] Separated field javascript into own function

### PR DESCRIPTION
**Pull request by tabacitu**
**Affects:** ```select2```, ```date_picker```, ```ckeditor``` and ALL other field types that have javascript;

## TLDR

Needed to do this to all field types:
Screenshot 2019-05-17 at 17 09 01
(call jQuery plugin on particular element, than on group selector; move jQuery plugin call into a method we can call any time we want)

## Context

In order to create the _matrix field type_ ( #771 ) and _drag-and-drop layout editor_ #1873 , we need to be able to programatically enable Javascript for a certain field. For example:
- in the matrix field type, you need to be able to tell the "date_picker" field to initialize js, after its put into the DOM;
- in the drag-and-drop layout field type, you need to be able to tell the "date_picker" field to initialize js, after its put into a modal where it can be edited;

A small rewrite of all field types is necessary for this, and it's a **breaking change**. Basically, each field type would need to:
- have a new attribute on the main element (input, select, textarea, whatever it is) that mentions what the javascript method is called (ex: ```data-init-function="bpFieldInitSelect2MultipleElement"```)
- move all the JS that enables the datpicker/select2/whatever inside a method with that name;
- the method will have an ```element``` parameter, which is the jQuery-wrapped node that had that event on it;
- make sure inside ```bpFieldInitXXXXX()```, javascript is initialized for one element (no ```each()```);

## Progress

Fields:
- [x] address_algolia
- [x] address_google
- [x] browse
- [x] browse_multiple
- [x] base64_image
- [x] checkbox
- [x] checklist
- [x] checklist_dependency
- [x] ckeditor
- [x] color
- [x] color_picker
- [x] custom_html
- [x] date
- [x] date_picker
- [x] date_range
- [x] datetime
- [x] datetime_picker
- [x] email
- [x] enum
- [x] hidden
- [x] icon_picker
- [x] image
- [x] month
- [x] number
- [x] page_or_link
- [x] password
- [x] radio
- [x] range
- [x] select (1-n relationship)
- [x] select_grouped
- [x] select2 (1-n relationship)
- [x] select_multiple (n-n relationship)
- [x] select2_multiple (n-n relationship)
- [x] select2_nested
- [x] select2_grouped
- [x] select_and_order
- [x] select_from_array
- [x] select2_from_array
- [x] select2_from_ajax
- [x] select2_from_ajax_multiple
- [x] simplemde
- [x] summernote
- [x] table
- [x] text
- [x] textarea
- [x] time
- [x] tinymce
- [x] upload
- [x] upload_multiple
- [x] url
- [x] video
- [x] view
- [x] week
- [x] wysiwyg

Todo:
- [x] use the Settings API to remember which type have been loaded, instead of the object on the model;

### CODE REVIEW

**Comment by tabacitu** [comment]
So apparently the current implementation with ```data-javascript-function-for-field-initialisation="bpFieldInitBrowseElement"``` isn't a perfect fix. Fields that plan to load other Backpack fields with Javascript, like the [matrix field type](https://github.com/Laravel-Backpack/CRUD/issues/771) and the [layout field type](https://github.com/Laravel-Backpack/CRUD/issues/1873) do NOT have the JS and CSS of the subsequent fields on page. That's because ```$crud->checkIfFieldIsFirstOfItsType($field)``` only looks for fields inside the _main form_, so will return ```false``` for any field that is used _outside_ the main form.

Will list a few solutions.

**Comment by tabacitu** [comment]
# Solution A

Implement [orchestral/asset](https://github.com/orchestral/asset) for asset management in Backpack/Base, and use that for fields too. 

Problem: It only supports JS and CSS _files_. That have an URL. You can't load JS and CSS blocks.

# Solution B

Add a general PHP array that remembers all fields types loaded on the page. So something like:

```php
    $loadedFieldTypes = [];

    public function getLoadedFieldTypes() {} // returns the array
    public function addLoadedFieldType($type) {} // add to the array
    public function fieldTypeLoaded($type) {} // returns true/false
```
